### PR TITLE
Experiment: use shallow constr hash to speedup hconstr hashconsing 

### PR DESCRIPTION
**Pull request by SkySkimmer**

### CODE REVIEW

**Comment by SkySkimmer** [comment]
with this the Time Defined in
~~~coq
Require Import Ltac2.Ltac2.

Require Import PrimInt63.

Inductive tree := Leaf (_:int) | Node (_:tree) (_:tree).

Ltac2 rec big node n :=
  if Int.equal n 0 then '(Leaf 0)
  else
    let n := Int.sub n 1 in
    let c := big node n in
    '($node $c $c).

Lemma foo : tree.
Proof.
  let c := big 'Node 20 in Std.exact_no_check c.
Time Defined.
~~~
is distributed between higher layers (close_proof), bytecode compilation and check-wf-univs.
With Qed to remove the bytecode compilation it's 80% close_proof, 20% check-wf-univs.

**Comment by SkySkimmer** [comment]
We probably shouldn't expect good results outside artificial examples as this relies on higher layers sending terms with good sharing to the kernel, but eg evar normalization breaks sharing easily:
~~~coq
Require Import Ltac2.Ltac2.

Require Import PrimInt63.

Inductive tree := Leaf (_:int) | Node (_:tree) (_:tree).

Ltac2 rec big node n :=
  if Int.equal n 0 then '(Leaf 0)
  else
    let n := Int.sub n 1 in
    let c := big node n in
    '($node $c $c).

Set Debug "hconstr,typeops".

Lemma foo : tree.
Proof.
  let c := big 'Node 22 in Std.exact_no_check c.
Time Qed.
(* 0.24s
   hconstr steps 26
   typeops steps 26 *)

Ltac2 rec big' n :=
  if Int.equal n 0 then exact (Leaf 0)
  else
    let n := Int.sub n 1 in
    refine
      (let c := '(_ :> tree) in
       '(Node $c $c));
    big' n.


Lemma foo' : tree.
Proof.
  big' 22.
Time Qed.
(* 1.8s
   hconstr steps 4194307
   qbindings = 24
   most_qcollisions = 19
   typeops steps 26 *)
~~~

Similarly on is_nat_500 from https://github.com/coq/coq/issues/18520
hconstr steps = 125254 (even if I remove the "limit collisions" commit (except the refcount fix which should really be another commit as nothing works without it), in which case qbindings = 999 and most_qcollisions = 496)
typeops steps = 1003

File: kernel/hConstr.ml [hunk 1]
```diff
@@ -256,34 +327,67 @@ let eq_leaf c c' = match kind c, c'.kind with
...
-      | _ -> 0, hash
+        uid, Hashset.Combine.combine uid hash, SList.defaultn (n-1) (SList.cons uid SList.empty)
+      | _ -> 0, hash, kind_unbound_rels kind
     in
-    { self; kind; hash; isRel; refcount = 1 }
+    { self; kind; hash; isRel; unbound_rels; refcount = 1 }
   in
-  match Tbl.find_opt tbl c with
+  match Tbl.find_opt (fst tbl) c with
   | Some c' -> c'.refcount <- c'.refcount + 1; c'
```
Comment (SkySkimmer) [inline kernel/hConstr.ml:389]: maybe should qadd here too
OTOH that may just increase the collision count

**Comment by ejgallego** [comment]
Very sadly I don't have time to read the code here, so just a note regarding hashes, sometimes you can make the hash of a DAG O(1) (amortized) by randomly picking children to hash; this needs to be properly seeded as to make the "randomness" deterministic. That could be a better O(1) strategy that the kind of bread-first `Hashbtl.hash_param` uses.

**Comment by SkySkimmer** [comment]
We do unconditionally recover sharing since https://github.com/coq/coq/pull/19160 so I'm not sure what you mean

**Comment by SkySkimmer** [comment]
The question here is can we exploit existing sharing when recovering sharing

**Comment by ejgallego** [comment]
> We do unconditionally recover sharing since #19160 so I'm not sure what you mean

I mean the lack of sharing F. Pottier pointed out.

**Comment by SkySkimmer** [comment]
If you mean https://github.com/coq/coq/issues/18520#issuecomment-2205905015 then AFAICT to_constr does produce a term with little sharing.

**Comment by SkySkimmer** [comment]
@coqbot bench

**Comment by SkySkimmer** [comment]
Don't know how to make this work well, discussion may continue in https://github.com/coq/coq/issues/18520

**Pull request closed without merging**